Add payment_settings to invoice

richardm-stripe opened the pull request:
r? @brandur-stripe
cc @stripe/api-libraries 

Adds support for `payment_settings` on the `Invoice` resource, from a recent [API Release](https://github.com/stripe/openapi/releases/tag/v16). (Skipping the "capability" part of that release for now, until that feature is fully released)

richardm-stripe commented on invoice.go:
```diff
@@ -29,6 +29,47 @@ const (
+// redirected to.
+type InvoicePaymentSettingsPaymentMethodOptionsBancontactPreferredLanguage string
+
+// List of values that InvoicePaymentSettingsPaymentMethodOptionsBancontactPreferredLanguage can take.
```
Not sure about this -- the general rule is "define enums for resources, but not for params" but we don't seem to be explicit about "language" enums anywhere else.

brandur-stripe replied:
Hmm, IMO we should probably just make it a string because (1) it's documented as string as opposed to enum in the API ref, (2) this could potentially have a lot of different values, (3) if the product team added a new value without the library I'm not sure that we'd be able to handle it that well, and (4) it looks like it's a string on payment intent, setup attempt, and charge.

Thoughts?

richardm-stripe commented on invoice.go:
```diff
@@ -29,6 +29,47 @@ const (
+// InvoicePaymentSettingsPaymentMethodType represents the payment method type to provide to the invoice's PaymentIntent.
+type InvoicePaymentSettingsPaymentMethodType string
+
+const (
```
```suggestion
// List of values that InvoicePaymentSettingsPaymentMethodType can take.
const (
```

richardm-stripe commented on invoice.go:
```diff
@@ -29,6 +29,47 @@ const (
+	InvoicePaymentSettingsPaymentMethodTypeGiropay            InvoicePaymentSettingsPaymentMethodType = "giropay"
+	InvoicePaymentSettingsPaymentMethodTypeIdeal              InvoicePaymentSettingsPaymentMethodType = "ideal"
+	InvoicePaymentSettingsPaymentMethodTypeSepaCreditTransfer InvoicePaymentSettingsPaymentMethodType = "sepa_credit_transfer"
+	InvoicePaymentSettingsPaymentMethodTypeSepaDebit          InvoicePaymentSettingsPaymentMethodType = "sepa_debit"
```
It's `SEPA` in account.go, but `Sepa` in `charge.go` and `payment_intent.go` I think. I figure it's better to be consistent with the "payment" resources than account.

remi-stripe replied:
That's fair, though that will increase the number of breaking changes once we codegen and fix this right? tbh I have no idea how we'll handle codegen-ing specific acronyms having to be in upper case 😓

brandur-stripe replied:
Arg lol. Yeah, that seems okay.

richardm-stripe commented on invoice.go:
```diff
@@ -29,6 +29,47 @@ const (
+type InvoicePaymentSettingsPaymentMethodType string
+
+const (
+	InvoicePaymentSettingsPaymentMethodTypeAchCreditTransfer  InvoicePaymentSettingsPaymentMethodType = "ach_credit_transfer"
```
Should be `ACH` but everywhere else it's just `Ach`.

richardm-stripe commented on invoice.go:
```diff
@@ -88,6 +129,31 @@ type InvoiceDiscountParams struct {
+// InvoicePaymentSettingsParams is the set of parameters allowed for the payment_settings on an invoice.
+type InvoicePaymentSettingsParams struct {
+	PaymentMethodOptions *InvoicePaymentSettingsPaymentMethodOptionsParams `form:"payment_method_options"`
+	PaymentMethodTypes   *[]*string                                        `form:"payment_method_types"`
```
```suggestion
	PaymentMethodTypes   []*string                                        `form:"payment_method_types"`
```

richardm-stripe commented on invoice.go:
```diff
@@ -29,6 +29,48 @@ const (
+
+// List of values that InvoicePaymentSettingsPaymentMethodType can take.
+const (
+	InvoicePaymentSettingsPaymentMethodTypeAchCreditTransfer  InvoicePaymentSettingsPaymentMethodType = "ach_credit_transfer"
```
Should be ACH, but it's Ach in other places and ACH does not exist anywhere.

remi-stripe commented on invoice.go:
```diff
@@ -88,6 +130,31 @@ type InvoiceDiscountParams struct {
+	PreferredLanguage *string `form:"preferred_language"`
+}
+
+// InvoicePaymentSettingsParams is the set of parameters allowed for card on payment_method_options on
```
Comment references the wrong type

brandur-stripe replied:
This is a bit nitpicky too, but I'd probably leave out things like `payment_method_options` just because the user will never directly interact with those API parameter names (i.e. they use Go symbols like `PaymentMethodOptions` instead).

I usually just keep the docs on these structs as simple as plausibly possible like "is the set of parameters for invoice payment settings". Not super useful I know, but pretty much no documentation will be unless we go for all-the-way API ref style docs (which might be possible with codegen).

remi-stripe commented on invoice.go:
```diff
@@ -88,6 +130,31 @@ type InvoiceDiscountParams struct {
+	RequestThreeDSecure *string `form:"request_three_d_secure"`
+}
+
+// InvoicePaymentSettingsParams is the set of parameters allowed for the payment_method_options on ther
```
Comment references the wrong type

remi-stripe commented on invoice.go:
```diff
@@ -88,6 +130,31 @@ type InvoiceDiscountParams struct {
+	Card       *InvoicePaymentSettingsPaymentMethodOptionsCardParams       `form:"card"`
+}
+
+// InvoicePaymentSettingsParams is the set of parameters allowed for the payment_settings on an invoice.
```
Comment references the wrong type

richardm-stripe replied:
Seems right to me

remi-stripe reviewed (commented):
nitpicky comments but deferring the review to brandur!

brandur-stripe commented:
A couple minor comments (and from Remi as well), but looks fine to me!

ptal @richardm-stripe

richardm-stripe commented on invoice.go:
```diff
@@ -29,6 +29,48 @@ const (
 	InvoiceBillingReasonUpcoming              InvoiceBillingReason = "upcoming"
 )
 
+// InvoicePaymentSettingsPaymentMethodOptionsBancontactPreferredLanguage is the
+// preferred language of the Bancontact authorization page that the customer is
+// redirected to.
+type InvoicePaymentSettingsPaymentMethodOptionsBancontactPreferredLanguage string
+
+// List of values that InvoicePaymentSettingsPaymentMethodOptionsBancontactPreferredLanguage can take.
+const (
+	InvoicePaymentSettingsPaymentMethodOptionsBancontactPreferredLanguageDE InvoicePaymentSettingsPaymentMethodOptionsBancontactPreferredLanguage = "de"
+	InvoicePaymentSettingsPaymentMethodOptionsBancontactPreferredLanguageEN InvoicePaymentSettingsPaymentMethodOptionsBancontactPreferredLanguage = "en"
+	InvoicePaymentSettingsPaymentMethodOptionsBancontactPreferredLanguageFR InvoicePaymentSettingsPaymentMethodOptionsBancontactPreferredLanguage = "fr"
+	InvoicePaymentSettingsPaymentMethodOptionsBancontactPreferredLanguageNL InvoicePaymentSettingsPaymentMethodOptionsBancontactPreferredLanguage = "nl"
+)
+
```
```suggestion
```

richardm-stripe commented on invoice.go:
```diff
@@ -88,6 +130,31 @@ type InvoiceDiscountParams struct {
+	PreferredLanguage *string `form:"preferred_language"`
+}
+
+// InvoicePaymentSettingsParams is the set of parameters allowed for card on payment_method_options on
```
```suggestion
// InvoicePaymentSettingsParams is the set of parameters allowed for invoice payment settings
```

richardm-stripe commented on invoice.go:
```diff
@@ -88,6 +130,31 @@ type InvoiceDiscountParams struct {
+	PreferredLanguage *string `form:"preferred_language"`
+}
+
+// InvoicePaymentSettingsParams is the set of parameters allowed for card on payment_method_options on
+// payment_settings on an invoice.
```
```suggestion
// InvoicePaymentSettingsPaymentMethodOptionsCardParams is the set of parameters allowed for
// payment method options when using the card payment method.
```

richardm-stripe commented on invoice.go:
```diff
@@ -88,6 +130,31 @@ type InvoiceDiscountParams struct {
+	RequestThreeDSecure *string `form:"request_three_d_secure"`
+}
+
+// InvoicePaymentSettingsParams is the set of parameters allowed for the payment_method_options on ther
+// payment_settings on an invoice.
```
```suggestion
// InvoicePaymentSettingsParams is the set of parameters allowed for specific payment methods
// on an invoice's payment settings.
```

richardm-stripe commented on invoice.go:
```diff
@@ -380,6 +449,30 @@ type InvoiceLineList struct {
+// InvoicePaymentSettingsPaymentMethodOptionsBancontact, if paying by `bancontact`, contains details about the Bancontact payment method options
+// to pass to the invoice's PaymentIntent.
+type InvoicePaymentSettingsPaymentMethodOptionsBancontact struct {
+	PreferredLanguage InvoicePaymentSettingsPaymentMethodOptionsBancontactPreferredLanguage `json:"preferred_language"`
```
```suggestion
	PreferredLanguage string `json:"preferred_language"`
```

richardm-stripe commented:
r? @brandur-stripe 
I
- Replaced the PreferredLanguage enum with a string.
- Kept it as Sepa, not SEPA.
- Kept as Ach, not ACH.
- Simplified/corrected docstrings for InvoicePaymentSettingsParams and its children.

brandur-stripe reviewed (approved):
Thanks Richard. LGTM.
